go-controller/pkg/ovn/policy.go

```diff
@@ -217,23 +220,31 @@ func deleteACLPortGroup(portGroupName, direction, priority, match, action string
 	return nil
 }
 
-func addToPortGroup(portGroup string, portInfo *lpInfo) error {
-	_, stderr, err := util.RunOVNNbctl("--if-exists", "remove",
-		"port_group", portGroup, "ports", portInfo.uuid, "--",
-		"add", "port_group", portGroup, "ports", portInfo.uuid)
-	if err != nil {
-		return fmt.Errorf("failed to add logicalPort %s to portGroup %s "+
-			"stderr: %q (%v)", portInfo.name, portGroup, stderr, err)
+func addToPortGroup(ovnNBClient goovn.Client, portGroupName string, portInfo *lpInfo) error {
+	cmd, err := ovnNBClient.PortGroupAddPort(portGroupName, portInfo.uuid)
+	if err == nil {
+		if err = ovnNBClient.Execute(cmd); err != nil {
+			return fmt.Errorf("execute error adding port %s to port group %s (%v)", portInfo.name, portGroupName, err)
+		}
+	} else if (err != goovn.ErrorExist) && (err != goovn.ErrorNotFound) {
```

Comment:
why do we need to care about --if-exists here (ErrorNotFound)? That should only before a remove right? I'm not sure why the old code was removing then adding, but your new code isn't doing that right?

with ovn-nbctl if i just keep adding the same port nothing happens, no error and no duplicate entries:
```
[root@ovn-control-plane ~]# ovn-nbctl list port_group
_uuid               : c6e24619-29af-41cc-a1c2-fcb87e940f8c
acls                : []
external_ids        : {name=clusterPortGroup}
name                : clusterPortGroup
ports               : [6d5e96a5-f822-440c-bd60-2fb92438ec3d, b7b66a64-9aa3-46b5-88b4-6c9bd047b2ec]

_uuid               : dd37c3e3-db2d-4843-a6a6-b0d723e4676a
acls                : []
external_ids        : {name=clusterRtrPortGroup}
name                : clusterRtrPortGroup
ports               : [6d5e96a5-f822-440c-bd60-2fb92438ec3d, ecb61096-56d1-4794-8531-5d76b4212802]
[root@ovn-control-plane ~]# ovn-nbctl add port_group c6e24619-29af-41cc-a1c2-fcb87e940f8c ports 6d5e96a5-f822-440c-bd60-2fb92438ec3d
[root@ovn-control-plane ~]# echo $?
0
[root@ovn-control-plane ~]# ovn-nbctl list port_group
_uuid               : c6e24619-29af-41cc-a1c2-fcb87e940f8c
acls                : []
external_ids        : {name=clusterPortGroup}
name                : clusterPortGroup
ports               : [6d5e96a5-f822-440c-bd60-2fb92438ec3d, b7b66a64-9aa3-46b5-88b4-6c9bd047b2ec]

_uuid               : dd37c3e3-db2d-4843-a6a6-b0d723e4676a
acls                : []
external_ids        : {name=clusterRtrPortGroup}
name                : clusterRtrPortGroup
ports               : [6d5e96a5-f822-440c-bd60-2fb92438ec3d, ecb61096-56d1-4794-8531-5d76b4212802]

```

Reply:
This is a good question. The way the old code removed the port and then added it looked kind of strange to me. My thought is that whoever implemented it that way did so to avoid an error when adding a port that already existed.  However, as you point out, you can keep adding a port without error so it wasn't really needed.  The go-ovn code returns an error if the port being added already exists in the port group, so we need to ignore that error if we don't care.

However, Now that I'm studying this more, I think the more interesting question is why am I ignoring goovn.ErrorNotFound?  Well, I did that because a test case was failing because it tried to add a port to port group clusterRtrPortGroup but clusterRtrPortGroup had not been created in the test case.  I was thinking that in the old code, the "--if-exists" was preventing the error from being returned.  However, there are two commands issued, and the "--if-exists" only applies to the first one.  I think that the real problem is that the fake ovn-nbctl doesn't return an error in this case but it should, and there's probably something wrong with the test case.  I need to look at this more.

Reply:
I did some more investigation, and the "sets up a shared gateway" ginkgo test had to be changed.  The old code used the fake exec to add ports to port groups that didn't exist in the test case.  Since the fake exec doesn't really execute, it didn't throw an error. Since the new go-ovn code actually gets executed to the point of creating objects in a cache, it detected the issue and returned an error.  I updated the test case to create the missing port groups before calling WatchNodes().  See master_test.go lines 1193-1195.  I then removed the exception for ErrorNotFound in addToPortGroup().